### PR TITLE
6504-slotNames-is-defined-in-ClassDescription-but-invoked-from-Behavior

### LINKED ISSUE #6504: slotNames is defined in ClassDescription but invoked from Behavior
Behavior >> definesSlotNamed: aString
	"Return true whether the receiver defines an instance variable named aString."
	
	^ self slotNames includes: aString

### PR DESCRIPTION
**Pull request by MarcusDenker**
- move #definesSlotNamed: to ClassDescription like al the other slot methods
- take this chance to redo the PR to move all extension methods of Slot class to the extended classes

fixes #6504

### CODE REVIEW

File: src/Collections-Strings/String.class.st [hunk 1]
```diff
@@ -320,6 +320,11 @@ String class >> newUppercasingTable [
 	^ String withAll: (Character allByteCharacters collect: [:c | c asUppercase])
 ]
 
+{ #category : #accessing }
+String class >> pseudovariablesNames [
+	^#('self' 'true' 'false' 'nil' 'thisContext' 'super')
+]
```
Comment (guillep) [inline src/Collections-Strings/String.class.st:326]: Why is this in String?

File: src/Collections-Strings/String.class.st [hunk 2]
```diff
@@ -526,6 +531,11 @@ String >> asCamelCase [
                        stream nextPutAll: sub capitalized]]
 ]
 
+{ #category : #variables }
+String >> asClassVariable [
+	^ self asSymbol asClassVariable
+]
+
```
Comment (guillep) [inline src/Collections-Strings/String.class.st:538]: Shouldn't this be an extension method?

File: src/Collections-Strings/String.class.st [hunk 3]
```diff
@@ -673,6 +683,22 @@ String >> asSignedInteger [
 	^ Integer readFrom: stream
 ]
 
+{ #category : #variables }
+String >> asSlot [
+	^ self asSymbol asSlot
+]
```
Comment (guillep) [inline src/Collections-Strings/String.class.st:689]: Should this be an extension method?

File: src/Collections-Strings/String.class.st [hunk 4]
```diff
@@ -673,6 +683,22 @@ String >> asSignedInteger [
...
+{ #category : #variables }
+String >> asSlotCollection [
+	"Parse as slots. Use space, tab and cr as separators
+	Example: 
+	  'a b' asSlotCollection --> {a => Slot. b => Slot}
+	"
+	| whitespaces |
+	whitespaces := ByteString withAll: { Character space . Character tab . Character cr }.
+	^(self substrings: whitespaces) collect: [ :substring | substring asSlot ]
+]
```
Comment (guillep) [inline src/Collections-Strings/String.class.st:700]: Same, Should this be an extension method?

File: src/Collections-Strings/String.class.st [hunk 5]
```diff
@@ -1857,6 +1883,14 @@ String >> isPatternVariable [
 	 ^self keywords anySatisfy: [:each | each first = $`]
 ]
 
+{ #category : #testing }
+String >> isPseudovariableName [
+	"Answer true if I am a pseudo-variable name.
+	#self isPseudovariableName -> true
+	"
+	^ self class pseudovariablesNames includes: self
+]
```
Comment (guillep) [inline src/Collections-Strings/String.class.st:1892]: Why is this in String? The only user is Slot. This should be moved to the Slot class...

```smalltalk
aSymbol isPseudovariableName ifTrue: [ 
		^ InvalidSlotName signalFor: aSymbol ].
```

```smalltalk
(self isPseudovariableName: aSymbol) ifTrue: [ 
		^ InvalidSlotName signalFor: aSymbol ].

```

I don't see the value of this being in `String`

File: src/Collections-Strings/Symbol.class.st [hunk 6]
```diff
@@ -410,12 +410,25 @@ Symbol >> = aSymbol [
 	^ super = aSymbol
 ]
 
+{ #category : #variables }
+Symbol >> => aVariable [
+	"If the slot we give as argument is not present in the image, we will get a nil. In that case we should throw an explicit error to the user saying a slot is missing."
+
+	aVariable ifNil: [ SlotNotFound signalForName: self ].
+	^ aVariable named: self
+]
```
Comment (guillep) [inline src/Collections-Strings/Symbol.class.st:419]: Why is this in `Symbol`? this should be at least in an extension method.

This is not good packaging...

File: src/Collections-Strings/Symbol.class.st [hunk 7]
```diff
@@ -410,12 +410,25 @@ Symbol >> = aSymbol [
...
 { #category : #converting }
 Symbol >> asAnnouncement [
 
 	^ self
 ]
 
+{ #category : #variables }
+Symbol >> asClassVariable [
+	^ ClassVariable named: self.
+]
```
Comment (guillep) [inline src/Collections-Strings/Symbol.class.st:430]: Same same

File: src/Collections-Strings/Symbol.class.st [hunk 8]
```diff
@@ -432,6 +445,11 @@ Symbol >> asMutator [
 	^ (self copyWith: $:) asSymbol
 ]
 
+{ #category : #variables }
+Symbol >> asSlot [
+	^ InstanceVariableSlot named: self.
+]
+
```
Comment (guillep) [inline src/Collections-Strings/Symbol.class.st:452]: Same.

File: src/Collections-Unordered/Dictionary.class.st [hunk 9]
```diff
@@ -391,6 +391,30 @@ Dictionary >> declare: key from: aDictionary [
...
+		self add: newGlobal.
+		].
+	(aDictionary includesKey: globalName)
+		ifTrue:  [
+			self add: ((aDictionary associationAt: globalName) primitiveChangeClassTo: ClassVariable new).
+			aDictionary removeKey: globalName]
+		ifFalse: [
+			self add: newGlobal]
+]
+
```
Comment (guillep) [inline src/Collections-Unordered/Dictionary.class.st:417]: Again. I don't see why this should be in Dictionary. This is no general use of dictionary.
Either we need a special "variable" dictionary. Or all this logic should be in the client.

File: src/System-Support/SystemDictionary.class.st [hunk 10]
```diff
@@ -170,6 +170,22 @@ SystemDictionary >> classOrTraitNamed: aString [
...
+
+	(self includesKey: key) ifTrue: [^ self].
+	(aDictionary includesKey: key)
+		ifTrue: 
+			[self add: ((aDictionary associationAt: key) primitiveChangeClassTo: GlobalVariable new).
+			aDictionary removeKey: key]
+		ifFalse: 
+			[self add: (GlobalVariable key: key)]
+]
+
```
Comment (guillep) [inline src/System-Support/SystemDictionary.class.st:188]: This looks pretty similar to `Dictionary >> declareVariable: newGlobal from: aDictionary` no?

**Review by guillep — CHANGES_REQUESTED** [review]
I think all changes related to `ClassDescription`, `Behavior`, closures are ok.

However, all the String related stuff is strange and pollutes the String API, for something not really useful.

**Comment by MarcusDenker** [comment]
Thanks! The only problem now I that I can not load the PR...  changes are done in an order that is bad (it remove a method that it then needs?)

I will redo it step by step with multiple PRs taking the review into account (I let it open for that)

**Comment by Ducasse** [comment]
Thank marcus. I agree with the comments of Guille. I try to avoid to have these as* on Symbol and String because it means that we cannot dispatch.

**Comment by MarcusDenker** [comment]
I added everything to my backlog to be done step by step, I close this as it will soon be conflicting with other merged PR.

**Pull request closed without merging**